include/pybind11/cast.h

```diff
@@ -604,41 +611,22 @@ template <> class type_caster<bool> {
     PYBIND11_TYPE_CASTER(bool, _("bool"));
 };
 
-template <> class type_caster<std::string> {
-public:
-    bool load(handle src, bool) {
-        object temp;
-        handle load_src = src;
-        if (!src) {
-            return false;
-        } else if (PyUnicode_Check(load_src.ptr())) {
-            temp = reinterpret_steal<object>(PyUnicode_AsUTF8String(load_src.ptr()));
-            if (!temp) { PyErr_Clear(); return false; }  // UnicodeEncodeError
-            load_src = temp;
-        }
-        char *buffer;
-        ssize_t length;
-        int err = PYBIND11_BYTES_AS_STRING_AND_SIZE(load_src.ptr(), &buffer, &length);
-        if (err == -1) { PyErr_Clear(); return false; }  // TypeError
-        value = std::string(buffer, (size_t) length);
-        success = true;
-        return true;
-    }
+// Helper class for UTF-{8,16,32} strings:
+template <typename CharT, class Traits, class Allocator>
+struct type_caster<std::basic_string<CharT, Traits, Allocator>, enable_if_t<is_std_char_type<CharT>::value>> {
+    static constexpr unsigned int UTF_N =
+        std::is_same<CharT, char>::value     ? 8 :
+        std::is_same<CharT, char16_t>::value ? 16 :
+        std::is_same<CharT, char32_t>::value ? 32 :
+        (sizeof(CharT) == 2 ? 16 : 32); /* std::wstring is UTF-16 on Windows, UTF-32 everywhere else */
```

Comment:
The `char16_t` and `char32_t` aren't guaranteed to be 2 and 4 bytes, only that they are at least 2/4 bytes (they are more like `std::uint_least16_t`).  They are guaranteed to be unique types, however, hence checking the actual types rather than just the sizes.